gemini/sample-apps/customer-search/functions/expense_prediction/main.py

```diff
@@ -0,0 +1,374 @@
+project_id = environ.get("PROJECT_ID")
+public_bucket = environ.get("PUBLIC_BUCKET")
+
+
+def create_graph(amount: list, category: list, date: list, cust_id: int) -> str:
+    """
+    Creates a graph of the predicted expenses.
+
+    Args:
+        amount (list): The amount of each expense.
+        category (list): The category of each expense.
+        date (list): The date of each expense.
+        cust_id (int): The customer ID.
+
+    Returns:
+        str: The public URL of the graph.
+    """
+
+    pio.templates.default = "plotly_white"
+
+    lst = []
+    for i in enumerate(amount):
+        if category[i] != "Healthcare":
+            lst.append(
+                [
+                    str(date[i].strftime("%b")) + " " + str(date[i].year),
+                    category[i],
+                    amount[i],
+                ]
+            )
+    df = pd.DataFrame(lst, columns=["Date", "Category", "Amount"])
+    df = df.sort_values(by="Date", ascending=False)
+
+    fig = px.line(
+        df,
+        x="Date",
+        y="Amount",
+        color="Category",
+        line_group="Category",
+        width=800,
+        height=600,
+    )
+    fig.update_yaxes(showgrid=True, gridwidth=1, zeroline=True)
+    fig.update_xaxes(showgrid=False, zeroline=True)
+
+    image_file_name = str(cust_id) + ".png"
+
+    image_dir = os.path.join(tempfile.gettempdir(), "prediction")
+    image_path = os.path.join(image_dir, image_file_name)
+
+    if not os.path.isdir(image_dir):
+        os.mkdir(image_dir)
+
+    fig.write_image(image_path)
+
+    client = storage.Client()
+    bucket = client.bucket(public_bucket)
+    cropped_image = bucket.blob(image_file_name)
+    cropped_image.upload_from_filename(image_path)
+    return cropped_image.public_url
+
+
+def create_aggregate_transaction_table() -> None:
+    """
+    Creates an aggregate transaction table.
+    """
+
+    client = bigquery.Client()
+    query = f"""
+    SELECT ac_id, DATE_TRUNC(date, MONTH) AS month_year,sub_category,category,\
+    sum(transaction_amount) as transaction_amount \
+    FROM `{project_id}.DummyBankDataset.AccountTransactions` \
+    WHERE debit_credit_indicator = 'Debit' \
+    GROUP BY ac_id,sub_category,category, month_year
+    """
+
+    job_config = bigquery.QueryJobConfig()
+
+    # Set configuration.query.destinationTable
+    destination_dataset = client.dataset("ExpensePrediction")
+    destination_table = destination_dataset.table("training_data")
+    job_config.destination = destination_table
+
+    # Set configuration.query.createDisposition
+    job_config.create_disposition = "CREATE_IF_NEEDED"
+
+    # Set configuration.query.writeDisposition
+    job_config.write_disposition = "WRITE_TRUNCATE"
+
+    # Start the query
+    job = client.query(query, job_config=job_config)
+
+    # Wait for the query to finish
+    job.result()
+
+
+def train_model() -> None:
+    """
+    Trains the expense prediction model.
+    """
+
+    client = bigquery.Client()
```

Comment:
Typically, you don't want to create a new client with each method call. This should be stateful